clippy_lints/src/methods.rs

```diff
@@ -1220,6 +1241,37 @@ fn lint_map_unwrap_or_else<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx hir
     }
 }
 
+/// lint use of `_.map_or(None, _)` for `Option`s
+fn lint_map_or_none<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx hir::Expr, map_or_args: &'tcx [hir::Expr]) {
+    // check if the first non-self argument to map_or() is None
+    let map_or_arg_is_none = if let hir::Expr_::ExprPath(ref qpath) = map_or_args[1].node {
+        match_qpath(qpath, &paths::OPTION_NONE)
+    } else {
+        false
+    };
+
+    if match_type(cx, cx.tables.expr_ty(&map_or_args[0]), &paths::OPTION) && map_or_arg_is_none {
+        // lint message
+        let msg = "called `map_or(None, f)` on an Option value. This can be done more directly by calling \
+                   `and_then(f)` instead";
+        let map_or_none_snippet = snippet(cx, map_or_args[1].span, "..");
+        let map_or_func_snippet = snippet(cx, map_or_args[2].span, "..");
+        let multiline = map_or_func_snippet.lines().count() > 1 || map_or_none_snippet.lines().count() > 1;
```

Comment:
Please get rid of the multiline check (I know all the others still do it, but they are rather old lints). Use `span_lint_and_then` + `span_suggestion` to produce the suggestion. Rustc will take care of the rest.